test/cpsubsystem/SessionAwareSemaphoreTest.js

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const chai = require('chai');
+chai.should();
+chai.use(require('chai-as-promised'));
+const expect = chai.expect;
+const fs = require('fs');
+const RC = require('./../RC');
+const {
+    Client,
+    IllegalStateError
+} = require('../../');
+
+describe('SessionAwareSemaphoreTest', function () {
+
+    this.timeout(30000);
+
+    let cluster;
+    let client;
+    let groupSeq = 0;
+
+    // we use a separate semaphore (and group) per each test to enforce test isolation
+    async function getSemaphore(initializeWith) {
+        const semaphore = await client.getCPSubsystem().getSemaphore('sessionaware@group' + groupSeq++);
+        if (initializeWith) {
+            await semaphore.init(initializeWith);
+        }
+        return semaphore;
+    }
+
+    before(async function () {
+        cluster = await RC.createCluster(null, fs.readFileSync(__dirname + '/hazelcast_cpsubsystem.xml', 'utf8'))
+        await Promise.all([
+            RC.startMember(cluster.id),
+            RC.startMember(cluster.id),
+            RC.startMember(cluster.id)
+        ]);
+        client = await Client.newHazelcastClient({ clusterName: cluster.id });
+    });
+
+    after(async function () {
+        await client.shutdown();
+        return RC.shutdownCluster(cluster.id);
+    });
+
+    it('release: should throw when not acquired', async function () {
+        const semaphore = await getSemaphore(3);
```

Comment:
It is more like a server-side implementation detail but what happens when I have two clients and client-a acquires 2 permits and client-b acquires 1 permits. What happens when client-a tries to release 3 permits?

Reply:
In case of the session-aware semaphore, the operation will fail as client-a doesn't hold enough permits. In case of session-less semaphore, it will succeed, as this implementation doesn't track ownership of permits.